File: app/src/main/java/makarglavanar/com/github/gitc/ui/issues/IssuesAdapter.java
package makarglavanar.com.github.gitc.ui.issues;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.devspark.robototextview.widget.RobotoTextView;

import java.util.ArrayList;
import java.util.List;

import makarglavanar.com.github.gitc.R;
import makarglavanar.com.github.gitc.entities.Issue;

public class IssuesAdapter extends RecyclerView.Adapter<IssuesAdapter.ViewHolder> {
    private List<Issue> issues;

    public IssuesAdapter() {
        this.issues = new ArrayList<>();
    }

    public void add(List<Issue> issues) {
        this.issues = issues;
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.item_issue_layout, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Issue issue = issues.get(position);
        holder.bind(issue);
    }

    @Override
    public int getItemCount() {
        return issues.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private Issue issue;
        private RobotoTextView title;
        private RobotoTextView userLoginView;
        private RobotoTextView createdDayView;

        public ViewHolder(View itemView) {
            super(itemView);
            title = (RobotoTextView) itemView.findViewById(R.id.titleView);
            userLoginView = (RobotoTextView) itemView.findViewById(R.id.userLogin);
            createdDayView = (RobotoTextView) itemView.findViewById(R.id.createdDay);
        }

        void bind(Issue issue) {
            this.issue = issue;
            title.setText(issue.getTitle());
            userLoginView.setText(issue.getUser().getLogin());
            createdDayView.setText(issue.getDate());
        }
    }
}

File: app/src/main/java/makarglavanar/com/github/gitc/ui/issues/issue_info/IssuesAdapter.kt
package makarglavanar.com.github.gitc.ui.issues.issue_info

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.item_issue_comment.view.*
import kotlinx.android.synthetic.main.item_issue_header_layout.view.*
import makarglavanar.com.github.gitc.R
import makarglavanar.com.github.gitc.entities.Issue
import makarglavanar.com.github.gitc.entities.IssueComment


class IssuesAdapter(
        val issue: Issue,
        val comments: List<IssueComment>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val header = 0
    private val comment = 1

    override fun getItemViewType(position: Int): Int {
        when (position) {
            header -> return header
            else -> return comment
        }
    }


    override fun getItemCount() = comments.size + 1

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            header -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_issue_header_layout, parent, false)
                return IssueHeaderViewHolder(view)
            }
            comment -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_issue_comment, parent, false)
                return CommentsViewHolder(view)
            }
            else -> throw RuntimeException("There is no type that matches the type $viewType + make sure your using types correctly")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is IssueHeaderViewHolder) holder.bind(issue)

        if (holder is CommentsViewHolder) {
            val comment = comments[position - 1]
            holder.bind(comment)
        }
    }


    class IssueHeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val titleView = itemView.issueTitleView
        private val bodyView = itemView.issueBodyView
        private val issueStateView = itemView.issueStateView
        private val userLoginView = itemView.issueOwnerLoginView

        fun bind(issue: Issue) {
            titleView.text = issue.title
            bodyView.text = issue.body
            issueStateView.text = issue.state
            userLoginView.text = issue.user.login
        }
    }


    class CommentsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val userLoginView = itemView.userLogin
        private val bodyView = itemView.commentBodyView
        private val createdDayView = itemView.createdDay
        private val userAvatarView = itemView.userAvatarView

        fun bind(comment: IssueComment) {
            userLoginView.text = comment.user.login
            bodyView.text = comment.body
            createdDayView.text = comment.created_at
        }
    }
}